docs/proposals/20190709-cluster-spec-crds.md

```diff
@@ -0,0 +1,329 @@
+    |                 |  Infrastructure update      |                             |
+    |                 |---------------------------> |                             |
+    |                 |                            +-+                            |
+    |                 |  +-----------------------------------------+              |    
+    |                 |  | IF Infrastructure.Status.Ready          |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  | Update Cluster Status   | |             |              |    
+    |                 |  | from Infrastructure Status             |              |    
+    |                 |<-+-------------------------| |             |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  | Set Cluster.Status.InfrastructureReady  |              |    
+    |                 |  | to "True"               | |             |              |    
+    |                 |<-+-------------------------| |             |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  +-----------------------------------------+              |
+    |                 |                            +-+                            |
+    |                 |                             |                             |
+```
+Figure 1: Cluster provisioning process
+
+Figure 1 presents the sequence of actions involved in provisioning a cluster, highlighting the coordination required between the CAPI cluster controller and the provider infrastructure controller. 
+
+The creation of the Cluster and provider infrastructure objects are independent events. It is expected that the cluster infrastructure object to be created before the cluster object and the reference be set to in the cluster object at creation time. 
+
+When a provider infrastructure object is created, the provider's controller will do nothing unless its owner reference is set to a cluster object. 
+
+When the cluster object is created, the cluster controller will retrieve the infrastructure object. If the object has not been seen before, it will start watching it. Also, if the object's owner is not set, it will set to the Cluster object. 
+
+When an infrastructure object is updated, the provider controller will check the owner reference. If it is set, it will retrieve the cluster object to obtain the required cluster specification and starts the provisioning process. When the process finishes, it sets the `Infrastructure.Status.Ready` to true.
+
+When the cluster controller detects the `Infrastructure.Status.Ready` is set to true, it updates the Cluster status with information from `Infrastructure.Status` (e.g. `APIEndpoint`) and sets `Cluster.Status.InfrastructureReady` to true.
```

Comment:
I think we need a stricter contract here around `APIEndpoints` to ensure this field is populated. We should define the required location in the infrastructure resource where the endpoints are to be found. The cluster controller copies the value from the infra resource to the cluster status field. WDYT?

Reply:
I agree. Will just InfrastructureRef->Status.Endpoints be sufficient? or are thinking about a structure were in the future we could add more fields?

Reply:
The status portion of the infrastructure object is the structure where we can add more fields over time. For now, I think it's sufficient to spec out the requirements around "ready" and "api endpoints".